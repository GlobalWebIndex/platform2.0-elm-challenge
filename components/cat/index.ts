export { CatActions } from './cat-actions'
export { CatList } from './cat-list'
export { CatView } from './cat-view'
