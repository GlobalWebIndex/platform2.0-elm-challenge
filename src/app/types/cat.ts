export type CatData = {
  id: string;
  url: string;
  width: number;
  height: number;
  breeds: CatBreed[];
  favourite: unknown;
};

export type CatBreed = {
  weight: unknown;
  id: string;
  name: string;
  temperament: string;
  origin: string;
  country_codes: string;
  country_code: string;
  life_span: string;
  wikipedia_url: string;
};
