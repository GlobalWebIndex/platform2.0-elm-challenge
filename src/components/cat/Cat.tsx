import React from 'react';

// Components
import { Button, Spinner } from '@blueprintjs/core';
// Store 
import { ICat, StatusEnum } from '../../constants/models';
import * as ComponentProps from '../../constants/props';
import useGlobal, { AppState, AssociatedActions, statelessActions } from '../../store/index';


export const Cat: React.SFC<ComponentProps.CatProps> = (props) => {

  const [favorites, aFavorites] = useGlobal((state: AppState) => state.favorites, (actions: AssociatedActions) => actions);
  let cat: ICat = { ...props.cat };
  let isFavorite: boolean = favorites.data.filter(o => o.image_id === cat.id).length !== 0;

  React.useEffect(() => {
    if (favorites.data.length === 0)
      aFavorites.fetchFavorites();
  });

  const toggleFavorite = async () => {
    isFavorite = favorites.data.filter(o => o.image_id === cat.id).length !== 0;
    if (!isFavorite) {
      await statelessActions.addFavorite(cat.id)
      aFavorites.fetchFavorites();
    }
    else {
      const favorite_id: number = favorites.data.filter(o => o.image_id === cat.id)[0].id
      await statelessActions.removeFavorite(favorite_id);
      aFavorites.fetchFavorites();
    }
  }

  return (
    <>
      <div className={props.isModal ? 'img-cat-details' : 'img-cat'}>
        <img src={cat.url} alt={'random Image' + cat.id} className="img-cat img-loading" />
        {props.isModal && cat.breeds && cat.breeds.length > 0 && <p><i>Breed name:</i> {cat.breeds[0].name}</p>}
        {props.isModal && (
          <Button
            icon="heart"
            onClick={toggleFavorite}
            disabled={favorites.status === StatusEnum.REQUESTED || favorites.status === StatusEnum.INITIAL}>
            {favorites.status === StatusEnum.REQUESTED && <Spinner size={Spinner.SIZE_SMALL}></Spinner>}
          </Button>
        )}
        {props.isModal && isFavorite && <i> In Favorites</i>}
      </div>
    </>
  );
}



