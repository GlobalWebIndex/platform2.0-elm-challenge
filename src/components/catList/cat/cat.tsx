import React from 'react';
import styles from './styles.scss';

type CatProps = {
  url: string;
};

export const Cat = ({url}: CatProps): JSX.Element => <img alt="Cat" className={styles.image} src={url} />;
