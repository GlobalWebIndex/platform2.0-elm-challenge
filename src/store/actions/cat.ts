import apiBase from "../config";
import { Store } from "use-global-hook";
import { AppState, AssociatedActions} from '../index'
import * as models from '../../constants/models'


export const fetchCats = (
    store: Store<AppState, AssociatedActions>,
    cats:  models.ApiListResult<models.ICat>
) => {

    cats = {...store.state.cats};
    cats.status = models.StatusEnum.REQUESTED;
    const rq = async () => {
        const res = await apiBase('/v1/images/search?limit=10');
        cats.data = res.data as models.ICat[];
        cats.status = models.StatusEnum.SUCCESS;
        store.setState( {...store.state, cats })
    }
    rq();
};
export const fetchMoreCats = (
    store: Store<AppState, AssociatedActions>,
    cats:  models.ApiListResult<models.ICat>
) => {
    cats = {...store.state.cats};
    cats.status = models.StatusEnum.REQUESTED;
    const rq = async () => {
        const res = await apiBase('/v1/images/search?limit=10');
        (cats.data).push( ...res.data as models.ICat[]);
        cats.status = models.StatusEnum.SUCCESS;
        store.setState( {...store.state, cats });
    }
    rq();
};

export const addCat = (
    store: Store<AppState, AssociatedActions>,
    cat:  models.ApiListResult<models.ICat>
) => {
    const newCats = {...store.state.cats, status: models.StatusEnum.SUCCESS}
    newCats.data.push(...cat.data);
    // store.setState( {...store.state.cats, newCats });
};


export const fetchCatsForBreed = (
    id: string
) => {
    const rq = async () => { 
        return await apiBase('/v1/images/search?limit=10&breed_id=' + id); 
    }
    return rq();
};

export const fetchCat = (
    id: string
) => {
    const rq = async () => { 
        return await apiBase('/v1/images/'+ id); 
    }
    return rq();
};