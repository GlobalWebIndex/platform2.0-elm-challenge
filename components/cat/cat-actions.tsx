'use client'

import { useOptimisticFavorite } from '@/hooks'
import { Bookmark, CommentBubble, Heart, HeartFilled } from '@/icons'
import { Cat } from '@/lib/types'
import { useRouter } from 'next/navigation'

interface CatProps {
  favoriteId?: string
  cat: Cat
}

export function CatActions({ cat, favoriteId }: CatProps) {
  const router = useRouter()
  const initialFavorite = !!favoriteId

  const { isFavorite, isPending, error, toggleFavorite } =
    useOptimisticFavorite(initialFavorite, favoriteId)

  const title = !isFavorite ? 'Add favorite' : 'Remove favorite'

  const handleFavorite = async () => {
    await toggleFavorite(cat)
    router.refresh()
  }

  return (
    <div className="mb-3 flex items-center justify-between">
      <div className="flex items-center gap-4">
        <button
          onClick={handleFavorite}
          title={title}
          aria-label={title}
          disabled={isPending}
          className="inline-flex size-6 items-center justify-center transition-opacity hover:text-neutral-500 hover:opacity-85"
        >
          {!isFavorite ? <Heart /> : <HeartFilled className="text-red-500" />}
        </button>
        <CommentBubble />
      </div>
      <Bookmark />
      {error && (
        <div className="cursor-default text-sm text-amber-600">{error}</div>
      )}
    </div>
  )
}
