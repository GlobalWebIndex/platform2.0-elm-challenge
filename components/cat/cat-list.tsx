'use client'

import { Button } from '@/components/button'
import { Image } from '@/components/image'
import { Spinner } from '@/components/spinner'
import { useApi } from '@/hooks'
import { Cat } from '@/lib/types'
import Link from 'next/link'

type CatListProps = {
  cats: Cat[]
}

export function CatList({ cats: initialCats }: CatListProps) {
  const {
    data: cats,
    loading,
    error,
    loadMore
  } = useApi<Cat>({
    url: '/api/cats',
    initialData: initialCats,
    paginated: true
  })

  if (error)
    return (
      <p className="text-sm font-medium">
        There was an error fetching the cats.
      </p>
    )

  if (!cats) return null

  return (
    <>
      <ul className="mb-8 grid w-full grid-cols-2 gap-0.5 sm:grid-cols-3 sm:gap-1">
        {cats.map(cat => (
          <li key={cat.id} className="transition-opacity hover:opacity-90">
            <Link href={`/cats/${cat.id}`}>
              <div className="aspect-square">
                <Image src={cat.url} fill alt="Cat image" />
              </div>
            </Link>
          </li>
        ))}
      </ul>
      <div className="mb-8 flex flex-col">
        {loading ? (
          <Spinner />
        ) : (
          <Button variant="secondary" onClick={loadMore}>
            Load more cats
          </Button>
        )}
      </div>
    </>
  )
}
