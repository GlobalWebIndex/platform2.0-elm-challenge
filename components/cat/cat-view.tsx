import { Avatar } from '@/components/avatar'
import { Image } from '@/components/image'
import { VerticalDots } from '@/icons'
import { Cat } from '@/lib/types'
import { cn } from '@/lib/utils'
import Link from 'next/link'
import { CatActions } from './cat-actions'

type CatViewProps = {
  cat: Cat
  breedCat: Cat
  favoriteId?: string
  className?: string
}

export function CatView({
  cat,
  breedCat,
  className,
  favoriteId
}: CatViewProps) {
  const { url: catSrc, breeds } = cat
  const { url: breedCatSrc } = breedCat

  if (!breeds) return

  const {
    id: breedId,
    name: breedName,
    origin,
    description,
    temperament
  } = breeds[0]

  const tags = temperament.split(', ')

  return (
    <div
      className={cn(
        'relative mx-auto flex h-full w-full max-w-[975px] flex-1 flex-col items-center bg-white shadow-lg sm:max-h-[600px]',
        className
      )}
    >
      <div className="flex h-full w-full flex-col sm:flex-row">
        <div className="flex aspect-square w-full flex-grow flex-col sm:min-w-[360px]">
          <Image src={catSrc} fill alt="Cat image" />
        </div>
        <div className="flex w-full flex-col sm:max-w-[420px]">
          <div className="flex items-center justify-between gap-2 border-b p-6">
            <div className="flex items-center gap-2">
              <Link href={`/breeds/${breedId}`} className="inline-flex">
                <Avatar src={breedCatSrc} />
              </Link>

              <div className="flex flex-col items-start">
                <div className="flex items-center gap-1">
                  <Link href={`/breeds/${breedId}`}>
                    <h2 className="text-sm font-semibold">{breedName}</h2>
                  </Link>
                  <span className="cursor-default select-none text-xs">â€¢</span>
                  <button className="inline-flex text-sm font-semibold text-[#0095f6]">
                    Follow
                  </button>
                </div>

                <p className="cursor-default text-xs">{origin}</p>
              </div>
            </div>
            <VerticalDots />
          </div>

          <div className="flex flex-col gap-2 p-6">
            <div className="line-clamp-4 text-sm">
              <span className="cursor-default font-semibold">{breedName}</span>{' '}
              {description}
            </div>
            <div className="inline-flex flex-wrap gap-1">
              {tags.map(tag => (
                <span
                  className="inline cursor-default text-xs text-[#0095f6]"
                  key={tag}
                >
                  # {tag.toLowerCase()}
                </span>
              ))}
            </div>
            <span className="cursor-default text-xs font-medium text-neutral-500">
              10w
            </span>
          </div>
          <div className="mt-auto border-t p-6">
            <CatActions cat={cat} favoriteId={favoriteId} />
            <div className="cursor-default text-sm font-semibold">
              38,289 likes
            </div>
            <div className="cursor-default text-xs text-neutral-500">
              June 18
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}
