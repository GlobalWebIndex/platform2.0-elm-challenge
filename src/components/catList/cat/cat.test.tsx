import React from 'react';
import {render} from '@testing-library/react';
import {Cat} from './cat';

describe('<Cat />', () => {
  it('should render the cat image', () => {
    const {container} = render(<Cat url="https://example.com/cat.jpg" />);

    expect(container).toMatchSnapshot();
  });
});
