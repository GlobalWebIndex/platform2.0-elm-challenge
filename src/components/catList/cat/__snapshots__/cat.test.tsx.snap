// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cat /> should render the cat image 1`] = `
<div>
  <img
    alt="Cat"
    class="image"
    src="https://example.com/cat.jpg"
  />
</div>
`;
